src/sync/deque.rs

```diff
@@ -0,0 +1,577 @@
+//! worker.push(1);
+//! worker.pop();
+//!
+//! // Stealers take data from the other end of the deque
+//! worker.push(1);
+//! stealer.steal();
+//!
+//! // Stealers can be cloned to have many stealers stealing in parallel
+//! worker.push(1);
+//! let stealer2 = stealer.clone();
+//! stealer2.steal();
+//! ```
+
+// FIXME: all atomic operations in this module use a SeqCst ordering. That is
+//      probably overkill
+
+use std::cell::UnsafeCell;
+use std::mem;
+use std::ptr;
+use std::sync::atomic::Ordering::SeqCst;
+use std::sync::atomic::AtomicIsize;
+use std::sync::Arc;
+
+use mem::epoch::{self, Atomic, Shared, Owned};
+
+// Once the queue is less than 1/K full, then it will be downsized. Note that
+// the deque requires that this number be less than 2.
+const K: isize = 4;
+
+// Minimum number of bits that a buffer size should be. No buffer will resize to
+// under this value, and all deques will initially contain a buffer of this
+// size.
+//
+// The size in question is 1 << MIN_BITS
+const MIN_BITS: u32 = 7;
+
+struct Deque<T> {
+    bottom: AtomicIsize,
+    top: AtomicIsize,
+    array: Atomic<Buffer<T>>,
+}
+
+// FIXME: can these constraints be relaxed?
+unsafe impl<T: Send + Sync> Send for Deque<T> {}
+unsafe impl<T: Send + Sync> Sync for Deque<T> {}
+
+/// Worker half of the work-stealing deque. This worker has exclusive access to
+/// one side of the deque, and uses `push` and `pop` method to manipulate it.
+///
+/// There may only be one worker per deque, and operations on the worker
+/// require mutable access to the worker itself.
+pub struct Worker<T> {
+    deque: Arc<Deque<T>>,
+}
+
+/// The stealing half of the work-stealing deque. Stealers have access to the
+/// opposite end of the deque from the worker, and they only have access to the
+/// `steal` method.
+///
+/// Stealers can be cloned to have more than one handle active at a time.
+pub struct Stealer<T> {
+    deque: Arc<Deque<T>>,
+}
+
+/// When stealing some data, this is an enumeration of the possible outcomes.
+#[derive(PartialEq, Eq, Debug)]
+pub enum Steal<T> {
+    /// The deque was empty at the time of stealing
+    Empty,
+    /// The stealer lost the race for stealing data, and a retry may return more
+    /// data.
+    Abort,
+    /// The stealer has successfully stolen some data.
+    Data(T),
+}
+
+// An internal buffer used by the chase-lev deque. This structure is actually
+// implemented as a circular buffer, and is used as the intermediate storage of
+// the data in the deque.
+//
+// This Vec<T> always has a length of 0, the backing buffer is just used by the
+// code below.
+struct Buffer<T> {
+    storage: UnsafeCell<Vec<T>>,
+    log_size: u32,
+}
+
+impl<T> Worker<T> {
+    /// Pushes data onto the front of this work queue.
+    pub fn push(&mut self, t: T) {
+        unsafe { self.deque.push(t) }
+    }
+
+    /// Pops data off the front of the work queue, returning `None` on an empty
+    /// queue.
+    pub fn pop(&mut self) -> Option<T> {
```

Comment:
We may want to call this `try_pop` to match what we're doing with other queues. (Granted, this data structure will probably never have a blocking variant...)